pkg/query/endpointset.go

```diff
@@ -238,15 +265,63 @@ func (c *endpointSetNodeCollector) Collect(ch chan<- prometheus.Metric) {
 	}
 }
 
-// EndpointSet maintains a set of active Thanos endpoints. It is backed up by Endpoint Specifications that are dynamically fetched on
-// every Update() call.
+// EndpointGroup groups common endpoints (having the same gRPC dial Opts and coming from the same DNS discovery) together.
+// It is backed up by set of *exthttp.Discoverer structs can give us addresses.
+type EndpointGroup struct {
+	d *httpconfig.Discoverer
+
+	dialOpts []grpc.DialOption
+}
+
+func NewEndpointGroup(d *httpconfig.Discoverer, dialOpts []grpc.DialOption) *EndpointGroup {
+	return &EndpointGroup{
+		d:        d,
+		dialOpts: dialOpts,
+	}
+}
+
+func removeDuplicateEndpointSpecs(logger log.Logger, duplicatedStores prometheus.Counter, specs []EndpointSpec) []EndpointSpec {
+	set := make(map[string]EndpointSpec)
+	for _, spec := range specs {
+		addr := spec.Addr()
+		if _, ok := set[addr]; ok {
+			level.Warn(logger).Log("msg", "Duplicate store address is provided", "addr", addr)
+			duplicatedStores.Inc()
+		}
+		set[addr] = spec
+	}
+	deduplicated := make([]EndpointSpec, 0, len(set))
+	for _, value := range set {
+		deduplicated = append(deduplicated, value)
+	}
+	return deduplicated
+}
+
+// Spec returns current set of endpoint specs.
+// Note that endpoint specifications return  can change dynamically. If some component is missing from the list, we assume it is no longer
+// accessible and we close gRPC client for it, unless it is strict.
+func (g *EndpointGroup) Spec() []EndpointSpec {
+	g.d.Endpoints() // TODO: ...
+
+	/*
+		Something like..
+		for _, dnsProvider := range []*dns.Provider{dnsStoreProvider, dnsRuleProvider, dnsExemplarProvider, dnsMetadataProvider, dnsTargetProvider} {
+					var tmpSpecs []query.EndpointSpec
+
+					for _, addr := range dnsProvider.Addresses() {
+						tmpSpecs = append(tmpSpecs, query.NewGRPCEndpointSpec(addr, false))
+					}
+					tmpSpecs = removeDuplicateEndpointSpecs(logger, duplicatedStores, tmpSpecs)
+					specs = append(specs, tmpSpecs...)
+				}
+	*/
+}
+
```

Comment:
write function that returns current set of endpt specs..
How?? what should be the inputs and return values :slightly_smiling_face: